applications/tari_testnet_miner/src/main.rs

```diff
@@ -53,18 +58,44 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     info!(target: LOG_TARGET, "Settings loaded");
 
-    // ToDo run logic
+    info!(target: LOG_TARGET, "Requesting new block");
 
     let mut base_node = TestNetMinerClient::connect(settings.base_node_address.unwrap())?;
     let request = tonic::Request::new(VoidParams {});
 
     let response = base_node.get_block(request).await?;
 
-    // ToDo miner logic
+    let mut miner = Miner::new(ConsensusRules::current());
+    let mut block = proc_block(response.into_inner());
 
-    // ToDo get private key logic
+    loop {
+        let mut line = String::new();
+        println!("Mining, press c to close and stop");
+        let b1 = std::io::stdin().read_line(&mut line).unwrap();
```

Comment:
You have to sit here hitting enter to mine blocks?

Let me direct your attention to various `ctrlc` crates on crates.io

Reply:
Futures version solves this, this was just dumping all functionality into a single thread.